README.md

```diff
@@ -29,28 +42,133 @@ running `sbcl` and evaluating `ql:*local-project-directories*`. Once
+- `:blas`: BLAS-backed code
+- `:lapack`: LAPACK-backed code
+- `:expokit`: expokit-backed code
+
+In most cases, one does not need to concern themselves with backends;
+MAGICL functionality should "just work" and dispatch to the
+appropriate backend. However, the programmer always has control, even
+dynamically in the program, of which backends should be used at a
+given time with the `magicl.backends:with-backends` macro. For instance,
+
+```
+(magicl.backends:with-backends (:blas :lisp)
+  ;; ... code ...
+  )
+```
+
+says that the code should be executed, always preferring
+`:blas`-accelerated functions, and using `:lisp`-implemented functions
+as a fall-back.
+
+```
+(magicl.backends:with-backends (:lisp)
+  ;; ... code ...
+  )
+```
+
+says to *only* use `:lisp`-implemented functions, even if other
+backends are loaded.
+
+
+The active backends can be found with the function
+`magicl.backends:active-backends`, which lists the backends to use in
+priority order.
+
+One can be even finer-grained than `with-backends`. Given a function
+`f` which has many backend implementations, one can get a specific
+implementation by using the function:
+
+```
+(magicl.backends:backend-implementation 'f :backend-name)
+```
+
+For instance
+
+```
+(magicl.backends:backend-implementation 'magicl:csd :lapack)
+```
+
+will give the implementation of the cosine-sine decomposition function
+in LAPACK. This can be called in exactly the same way `magicl:csd` can
+be called.
+
+In `backend-implementation`, if both the function name and the backend
+name are (quoted) constants, this will be looked up at compile-time,
+which is useful for writing efficient code that does not dispatch. But
+note that by doing this, `with-backends` will not be respected.
+
 ## Testing MAGICL
 
 You can run the MAGICL tests from your Lisp REPL with:
 
 ```
-(ql:quickload :magicl-tests)
 (asdf:test-system :magicl)
 ```
 
+You currently need all of the extensions working for the tests to run.
+
 ## High-level Interface
 
-See [high-level doc](doc/high-level.md).
+See the [high-level doc](doc/high-level.md) for an extensive discussion
+and comparison of MAGICL functions with those of MATLAB and NumPy.
+
+## Developer's Guide: How to Add New Functions
+
+See the [developer how-to](doc/dev-how-to.md) to understand how to add
+new functionality to MAGICL.
 
 ## Showing Available Functions
 
-Some distributions of a library don't actually provide all of the functions of the reference BLAS and LAPACK. One can look at a summary of available and unavailable functions with the function `magicl:print-availability-report`. By default, it will show all functions and their availability. There are three arguments to fine-tune this behavior:
+Some distributions of a library don't actually provide all of the
+functions of the reference BLAS and LAPACK. One can look at a summary
+of available and unavailable functions with the function
+`magicl:print-availability-report`. By default, it will show all
```

Comment:
This is not currently exported from `magicl`, but is exported from `magicl.foreign-libraries`.

Reply:
good point; I added it as a re-export from MAGICL.